cmake/FindLLVM_config.cmake

```diff
@@ -0,0 +1,199 @@
+#  LLVM_MODULE_LIBS         - list of llvm libs for working with modules.
+#  LLVM_PACKAGE_VERSION     - Installed version.
+#  LLVM_TOOLS_BINARY_DIR    - Where to find llvm-* programs
+
+if(NOT LLVM_FOUND)
+  set(LLVM_KNOWN_MAJOR_VERSIONS
+      15
+      14
+      13
+      12
+      11
+      10
+      9
+      8
+      7
+      6.0
+      5.0
+      4.0
+      3.9
+      3.8)
+  foreach(ver ${LLVM_KNOWN_MAJOR_VERSIONS})
+    find_package(LLVM ${ver} QUIET)
+    if(LLVM_FOUND)
+      # Call the main findLLVM.cmake to check LLVM setup and set some flags such as:
+      #   LLVM_LIBRARY_DIR, LLVM_INCLUDE_DIR, LLVM_TOOLS_BINARY_DIR, LLVM_VERSION_MAJOR, LLVM_PACKAGE_VERSION
+      if(LLVM_FIND_VERBOSE)
+        find_package(LLVM ${ver} CONFIG)
+      else(LLVM_FIND_VERBOSE)
+        find_package(LLVM ${ver} CONFIG QUIET)
+      endif(LLVM_FIND_VERBOSE)
+      set(LLVM_config_FOUND TRUE)
+      break()
+    endif(LLVM_FOUND)
+  endforeach(ver ${LLVM_KNOWN_VERSIONS})
+endif(NOT LLVM_FOUND)
+
+if(NOT LLVM_CONFIG_EXECUTABLE)
+  find_program(
+    LLVM_CONFIG_EXECUTABLE
+    NAMES "llvm-config-${LLVM_VERSION_MAJOR}" "llvm-config"
+    DOC "llvm-config executable"
+    PATHS ${LLVM_TOOLS_BINARY_DIR})
+endif(NOT LLVM_CONFIG_EXECUTABLE)
+
+if(LLVM_CONFIG_EXECUTABLE)
+  message(STATUS "Found components for LLVM")
+  message(STATUS "llvm-config ........ = ${LLVM_CONFIG_EXECUTABLE}")
+else(LLVM_CONFIG_EXECUTABLE)
+  message(STATUS "llvm-config ........ = NOT FOUND")
+  message(WARNING "Could NOT find LLVM config executable")
+endif(LLVM_CONFIG_EXECUTABLE)
+
+execute_process(
+  COMMAND ${LLVM_CONFIG_EXECUTABLE} --cxxflags
+  OUTPUT_VARIABLE LLVM_CFLAGS
+  OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+execute_process(
+  COMMAND ${LLVM_CONFIG_EXECUTABLE} --ldflags
+  OUTPUT_VARIABLE LLVM_LFLAGS
+  OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+execute_process(
+  COMMAND ${LLVM_CONFIG_EXECUTABLE} --libs
+  OUTPUT_VARIABLE LLVM_MODULE_LIBS
+  OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+execute_process(
+  COMMAND ${LLVM_CONFIG_EXECUTABLE} --libfiles
+  OUTPUT_VARIABLE LLVM_MODULE_LIBFILES
+  OUTPUT_STRIP_TRAILING_WHITESPACE)
+separate_arguments(LLVM_MODULE_LIBFILES)
+
+# This should never happen
+if(NOT LLVM_TOOLS_BINARY_DIR)
+  execute_process(
+    COMMAND ${LLVM_CONFIG_EXECUTABLE} --bindir
+    OUTPUT_VARIABLE LLVM_TOOLS_BINARY_DIR
+    OUTPUT_STRIP_TRAILING_WHITESPACE)
+endif(NOT LLVM_TOOLS_BINARY_DIR)
+
+if(NOT CLANG_EXE_NAME)
+  find_program(
+    CLANG_EXE_FULLPATH
+    NAMES "clang-${LLVM_VERSION_MAJOR}" "clang"
+    DOC "clang executable"
+    PATHS ${LLVM_TOOLS_BINARY_DIR})
+  get_filename_component(CLANG_EXE_NAME ${CLANG_EXE_FULLPATH} NAME)
+endif(NOT CLANG_EXE_NAME)
+if(NOT OPT_EXE_NAME)
+  find_program(
+    OPT_EXE_FULLPATH
+    NAMES "opt-${LLVM_VERSION_MAJOR}" "opt"
+    DOC "opt executable"
+    PATHS ${LLVM_TOOLS_BINARY_DIR})
+  get_filename_component(OPT_EXE_NAME ${OPT_EXE_FULLPATH} NAME)
+endif(NOT OPT_EXE_NAME)
+
+set(LLVM_SYSTEM pthread z tinfo)
```

Comment:
This list of libraries depends on how LLVM is built and therefore cannot be hardcoded. Use `llvm-config --system-libs` to get the list instead of hard-coding it.

Reply:
I didn't know where do those settings came out from and why they were needed, thank you for the explaination!